crates/cli/src/commands.rs

```diff
@@ -73,9 +74,121 @@ pub struct CompileAndBuildCommandOpts {
     pub no_source_compression: bool,
 }
 
+#[derive(Debug, Parser)]
+pub struct BuildCommandOpts {
+    #[arg(value_name = "INPUT", required = true)]
+    /// Path of the JavaScript input file.
+    pub input: PathBuf,
+
+    #[arg(short, default_value = "index.wasm")]
+    /// Desired path of the WebAssembly output file.
+    pub output: PathBuf,
+
+    #[arg(
+        short = 'C',
+        long_help = "Available codegen options:
+-C dynamic[=y|n] -- Creates a smaller module that requires a dynamically linked QuickJS provider Wasm module to execute (see `emit-provider` command).
+-C wit=path -- Optional path to WIT file describing exported functions. Only supports function exports with no arguments and no return values.
+-C wit-world=val -- Optional WIT world name for WIT file. Must be specified if WIT is file path is specified.
+-C no-source-compression[=y|n] -- Disable source code compression, which reduces compile time at the expense of generating larger WebAssembly files.
+    "
+    )]
+    /// Codegen options.
+    pub codegen: Vec<CodegenOption>,
+}
+
 #[derive(Debug, Parser)]
 pub struct EmitProviderCommandOpts {
     #[structopt(short, long)]
     /// Output path for the provider binary (default is stdout).
     pub out: Option<PathBuf>,
 }
+
+#[derive(Clone, Debug, Parser)]
+pub struct CodegenOptionGroup {
+    /// Creates a smaller module that requires a dynamically linked QuickJS provider Wasm
+    /// module to execute (see `emit-provider` command).
+    pub dynamic: bool,
+    /// Optional path to WIT file describing exported functions.
+    /// Only supports function exports with no arguments and no return values.
+    pub wit: Option<PathBuf>,
+    /// Optional path to WIT file describing exported functions.
+    /// Only supports function exports with no arguments and no return values.
+    pub wit_world: Option<String>,
+    /// Disable source code compression, which reduces compile time at the expense of generating larger WebAssembly files.
+    pub no_source_compression: bool,
+}
+
+#[derive(Clone, Debug)]
+pub enum CodegenOption {
+    /// Creates a smaller module that requires a dynamically linked QuickJS provider Wasm
+    /// module to execute (see `emit-provider` command).
+    Dynamic(bool),
+    /// Optional path to WIT file describing exported functions.
+    /// Only supports function exports with no arguments and no return values.
+    Wit(PathBuf),
+    /// Optional path to WIT file describing exported functions.
+    /// Only supports function exports with no arguments and no return values.
+    WitWorld(String),
+    /// Disable source code compression, which reduces compile time at the expense of generating larger WebAssembly files.
+    NoSourceCompression(bool),
+}
+
+impl FromStr for CodegenOption {
```

Comment:
We had briefly discussed using https://docs.rs/clap/latest/clap/builder/trait.ValueParserFactory.html -- although the current approach seems to work well enough for the current PR. We can always iterate and refactor later.

Reply:
I should've mentioned in the description, I looked briefly at using it. My understanding is it works well if you need to keep track additional state outside of parsing the string to whatever type you're parsing it too. But the tradeoff is it forces you to define an additional type for the parser for the value, which is one more type to define. Ultimately since we have to collect a vector of _something_, I figured it would be easier to just write the state we need to collect to that _something_. That said, if we want to ensure there are no duplicate keys present and error out in the option collection phase, then having a custom parser would make that possible. Right now if an option is specified more than once, it'll take the last value as the one to use.

Reply:
>  That said, if we want to ensure there are no duplicate keys present and error out in the option collection phase, then having a custom parser would make that possible

In the long run, I think we'd want to ensure this. 

A custom parser also gives you more control over how to define/manage your group options and group help (e.g., you can have the help be automatically generated from the options without having to manually update the help string manually every time, which might be error prone).

I think it's fine to leave it as it for this first iteration, we can always refactor later.